src/include/txn.i

```diff
@@ -511,30 +519,63 @@ __wt_txn_visible(
  * __wt_txn_upd_visible --
  *	Can the current transaction see the given update.
  */
-static inline bool
+static inline WT_VISIBLE_TYPE
 __wt_txn_upd_visible(WT_SESSION_IMPL *session, WT_UPDATE *upd)
 {
-	return (__wt_txn_visible(session,
-	    upd->txnid, WT_TIMESTAMP_NULL(&upd->timestamp)));
+	uint8_t upd_state;
+	bool upd_visible;
+
+	for (;;) {
+		upd_state = upd->state;
+		if (upd_state == WT_UPDATE_STATE_LOCKED) {
+			/* Commit is in progress, yield and try again. */
+			__wt_yield();
+			continue;
+		}
+
+		upd_visible = __wt_txn_visible(session, upd->txnid,
+		    WT_TIMESTAMP_NULL(&upd->timestamp));
+		/*
+		 * Update state changed from beneath during txn visibility
+		 * check, try again.
+		 */
+		if (upd->state != upd_state)
+			continue;
```

Comment:
the state can change after the txn visibility check, before line 542, and in that case we will continue. Also the state can change after we have done this check. Can you explain further in the comment, why this is okay.